File: works/1991.py
import sys


class Node:
    def __init__(self, root, left, right):
        self.root = root
        self.left = left
        self.right = right


pre = []
inorder = []
post = []
map = {}


def PreOrder(root):
    if root == '.': return
    pre.append(root)
    PreOrder(map[root].left)
    PreOrder(map[root].right)


def PostOrder(root):
    if root == '.': return
    PostOrder(map[root].left)
    PostOrder(map[root].right)
    post.append(root)


def InOrder(root):
    if root == '.': return
    InOrder(map[root].left)
    inorder.append(root)
    InOrder(map[root].right)


n = int(input())
for i in range(0,n):
    _s = input()
    map[_s[0]] = Node(_s[0], _s[2], _s[4])

PreOrder('A')
PostOrder('A')
InOrder('A')

for x in pre: print(x, end='')
print()
for x in inorder: print(x, end='')
print()
for x in post: print(x, end='')

File: baekjoon/1991.cpp
#include <iostream>
#include <vector>
using namespace std;

int N;
vector<char> pre,in,post;

struct tree {
    char node;
    tree *LChild;
    tree *RChild;
};

void PreOrder(tree *root) {
    if(root == NULL) return;
    pre.push_back(root->node);
    PreOrder(root->LChild);
    PreOrder(root->RChild);
}

void InOrder(tree *root) {
    if(root == NULL) return;
    InOrder(root->LChild);
    in.push_back(root->node);
    InOrder(root->RChild);
}

void PostOrder(tree *root) {
    if(root == NULL) return;
    PostOrder(root->LChild);
    PostOrder(root->RChild);
    post.push_back(root->node);
}

int main() {
    vector<tree> v;
    for(char i = 'A'; i <= 'Z'; i++) {
        v.push_back({i, NULL, NULL});
    }

    cin >> N;
    for(int i = 0; i < N; i++) {
        char parent, left, right;
        cin >> parent >> left >> right;

        if(left != '.') {
            v[(int)(parent-'A')].LChild = &v[(int)(left-'A')];
        }
        if(right != '.') {
            v[(int)(parent-'A')].RChild = &v[(int)(right-'A')];
        }
    }

    PreOrder(&v[0]);
    InOrder(&v[0]);
    PostOrder(&v[0]);

    for(auto x : pre) cout << x; cout << '\n';
    for(auto x : in) cout << x; cout << '\n';
    for(auto x : post) cout << x; cout << '\n';

}